### PR TITLE
optimizes icon2html() for icon files known to be in the rsc at compile time

### PR DESCRIPTION
**Pull request by Kylerace**
## About The Pull Request
icon2html() creates assets for the asset cache system in order to transport the icons of the examined objects to clients, the process currently involves fcopy_rsc()'ing the icon memory object into the dynamic resource cache, md5() the contents of that, then creating the asset cache item fcopy_rsc()'s the icon again, then calls md5asfile() to guarantee correct hashing of the file contents by fcopy()'ing the rsc reference of the icon to a temporary file, then md5'ing the contents of that, then deleting the file. all so that if we ever call icon2html() on identical icon parameters we wont have to duplicate work sending things to clients. 

this is 6 file io calls by my count (im pretty sure the second fcopy_rsc() call of the icon doesnt do much work since its already put into the rsc, so at least 5), and since its blocking for our code it takes a long time of processing where the cpu is just waiting for io to return. thus causing md5asfile() to be one of the highest overtiming procs on live Screenshot_2041

now the two fcopy_rsc()'s could be collapsed into one (though it wont do much), but in order for the system to work the hash we use needs to be correct. which is why the fcopy() is used (md5'ing the rsc reference is unreliable when the icon being put into the rsc was given invalid data). but a weird quirk about icon objects is that you can find their dmi file path if and only if they are either: compile-time rsc references, or are unchanged subsets of compile-time rsc files. it just so happens that atom icon vars act as the first thing. if we can derive a dmi file path from the object given to us in icon2html(), we know that we can md5 the rsc reference correctly without any errors, because it directly represents a known dmi file (otherwise we couldnt get the dmi path)

thus, we can get rid of the most expensive parts of asset generation if we know we can rely on the md5 of the rsc file, which we can the vast majority of the time, because most examined objects use compile time icons which are md5'able

## Why It's Good For The Game
before i work on doing any rust stuff, this can help cut down on overtime. next up is getFlatIcon()
brings icon2html cost from this:

|proc|self|total|realtime|overtime|# calls|
|----|---|-----|--------|--------|--------|
|/proc/icon2html|0.003|0.646|0.646|0.001|69|

to this 

|proc|self|total|realtime|overtime|# calls|
|----|---|-----|--------|--------|--------|
|/proc/icon2html|0.004|0.075|0.075|0.000|72|

measured by going around and examining a bunch of objects once (otherwise it would be caching the cost). so going from ~9ms of total cpu per call to ~1 ms

also for rare cases where we cant derive a dmi filepath from the given icon, we will no longer use the md5() of the fcopy_rsc()'d rsc reference, which makes the key more accurate

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Fucking 10x reduction pog

**Comment by LemonInTheDark** [comment]
It's also worth noting this makes the stat panel object tab thing like 10x better for free

**Comment by LemonInTheDark** [comment]
fuck

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -1127,14 +1141,90 @@ GLOBAL_LIST_INIT(freon_color_matrix, list("#2E5E69", "#60A2A8", "#A1AFB1", rgb(0
...
+
+		icon_path = rsc_ref_string
+
+	else if(isatom(icon) || istype(icon, /image) || istype(icon, /mutable_appearance))
+		var/atom/atom_icon = icon
+		icon = atom_icon.icon
+		//atom icons compiled in from 'icons/path/to/dmi_file.dmi' are weird and not really icon objects that you generate with icon().
+		//if theyre unchanged dmi's then they're stringifiable to "icons/path/to/dmi_file.dmi"
+
+		return get_icon_dmi_path(icon)
```
Comment (MrStonedOne) [inline code/__HELPERS/icons.dm:1201]: if you did this at the top of the proc in its own if chain you could skip the recursive call

Comment (Kylerace) [reply code/__HELPERS/icons.dm]: good idea

**Pull request merged**